frame/system/src/extras.rs

```diff
@@ -0,0 +1,109 @@
+/// Hash for runtime.
+pub type HashFor<R> = <R as crate::Trait>::Hash;
+/// AccountId type for runtime.
+pub type AccountIdFor<R> = <R as crate::Trait>::AccountId;
+
+///
+/// Strongly typed interface for the `SignedExtensions` that should be included in extrinsics
+/// for them to valid for your runtime.
+///
+/// ```rust,ignore
+/// use runtime::{Runtime, SignedExtra};
+/// use frame_system::extras::IndexFor;
+///
+/// #[derive(Default)]
+/// struct ExtraParams {
+///		
+/// }
+///
+/// impl RuntimeAdapter for Runtime {
+///     type Extra = SignedExtra;
+///
+///		type Params = ExtraParams;
+///	
+///		fn extension_params() -> Self::Params {
+///			Default::default()
+/// 	}
+///		
+///     fn construct_extras(params: ExtraParams) -> Result<SignedExtensionData<Self::Extra>, &'static str> {
+///         // take the biggest period possible.
+///         let period = BlockHashCount::get()
+///             .checked_next_power_of_two()
+///             .map(|c| c / 2)
+///             .unwrap_or(2) as u64;
+///         let current_block = System::block_number()
+///             .saturated_into::<u64>()
+///             // The `System::block_number` is initialized with `n+1`,
+///             // so the actual block number is `n`.
+///             .saturating_sub(1);
+///
+///         (
+///             frame_system::CheckSpecVersion::new(),
+///             frame_system::CheckTxVersion::new(),
+///             frame_system::CheckGenesis::new(),
+///             frame_system::CheckEra::from(generic::Era::mortal(period, current_block)),
+///             frame_system::CheckNonce::from(index),
+///             frame_system::CheckWeight::new(),
+///         )
+///     }
+/// }
+///
+/// ```
+///
+pub trait SignedExtensionProvider: crate::Trait {
+    /// Concrete SignedExtension type.
+	type Extra: SignedExtension;
+	/// Concrete type for params used to construct the `SignedExtension`-Data
+	type Params: SystemExtraParams<Self>;
+
+	/// retrieve an instance of the input object.
+	fn extension_params() -> Self::Params;
+
+    /// construct extras and optionally additional_signed data for inclusion in extrinsics.
+    fn construct_extras(input: Self::Params) -> Result<SignedExtensionData<Self::Extra>, &'static str>;
+}
+
+/// extras that should be included in extrinsics,
+/// additional data is provided for call sites that don't have access to storage.
+pub struct SignedExtensionData<S: SignedExtension> {
+	/// signed extras
+	pub extra: S,
+	/// additional data for the signed extras.
+	pub additional: Option<S::AdditionalSigned>,
+}
+
+/// used internally by substrate to set extras for inclusion in 
+/// `SignedExtensionProvider`
+pub trait SystemExtraParams<T: crate::Trait> {
+	/// sets the nonce
+	fn set_nonce(&mut self, index: T::Index);
+	/// sets the nonce
+	fn set_era(&mut self, era: Era);
+	/// sets the prior block hash
+	fn set_prior_block_hash(&mut self, hash: T::Hash);
```

Comment:
What is the prior hash?

Reply:
this is the additional signed data for `CheckEra`, the hash of the block that starts the era. will rename it and add docs so that it's clearer